libs/elasticsearch-nio/src/main/java/org/elasticsearch/nio/EventHandler.java

```diff
@@ -20,15 +20,171 @@
+     *
+     * @param context that accepting a connection
+     * @param exception that occurred
+     */
+    protected void acceptException(ServerChannelContext context, Exception exception) {
+        context.handleException(exception);
+    }
+
+    /**
+     * This method is called when a NioChannel is being registered with the selector. It should
+     * only be called once per channel.
+     *
+     * @param context that was registered
+     */
+    protected void handleRegistration(ChannelContext<?> context) throws IOException {
+        context.register();
+        SelectionKey selectionKey = context.getSelectionKey();
+        selectionKey.attach(context);
+        if (context instanceof SocketChannelContext) {
+            if (((SocketChannelContext) context).readyForFlush()) {
+                SelectionKeyUtils.setConnectReadAndWriteInterested(context.getSelectionKey());
+            } else {
+                SelectionKeyUtils.setConnectAndReadInterested(context.getSelectionKey());
+            }
+        } else {
+            assert context instanceof ServerChannelContext : "If not SocketChannelContext the context must be a ServerChannelContext";
+            SelectionKeyUtils.setAcceptInterested(context.getSelectionKey());
+        }
+    }
+
+    /**
+     * This method is called when an attempt to register a channel throws an exception.
+     *
+     * @param context that was registered
+     * @param exception that occurred
+     */
+    protected void registrationException(ChannelContext<?> context, Exception exception) {
+        context.handleException(exception);
+    }
+
+    /**
+     * This method is called when a NioSocketChannel has just been accepted or if it has receive an
+     * OP_CONNECT event.
+     *
+     * @param context that was registered
+     */
+    protected void handleConnect(SocketChannelContext context) throws IOException {
+        if (context.connect()) {
+            SelectionKeyUtils.removeConnectInterested(context.getSelectionKey());
+        }
+    }
+
+    /**
+     * This method is called when an attempt to connect a channel throws an exception.
+     *
+     * @param context that was connecting
+     * @param exception that occurred
+     */
+    protected void connectException(SocketChannelContext context, Exception exception) {
+        context.handleException(exception);
+    }
+
+    /**
+     * This method is called when a channel signals it is ready for be read. All of the read logic should
+     * occur in this call.
+     *
+     * @param context that can be read
+     */
+    protected void handleRead(SocketChannelContext context) throws IOException {
+        context.read();
+    }
+
+    /**
+     * This method is called when an attempt to read from a channel throws an exception.
+     *
+     * @param context that was being read
+     * @param exception that occurred
+     */
+    protected void readException(SocketChannelContext context, Exception exception) {
```

Comment:
do we really need to make all these methods protected of can they be pkg private and final? I know some of them are used in a test subclass